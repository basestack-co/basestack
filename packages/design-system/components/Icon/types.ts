import { SpaceProps } from "styled-system";

export type Size = "xxLarge" | "xLarge" | "large" | "medium" | "small";

export interface IconProps extends SpaceProps {
  /**
   * Icon name
   */
  icon: string;
  /**
   * Icon size
   */
  size?: Size;
  /**
   * Icon color
   */
  color?: string;
  /**
   * Icon muted color
   */
  muted?: boolean;
}

export type CustomIconNames = "github";

export interface CustomIconProps {
  size: number;
  color?: string;
}

export interface CustomIconCompProps extends CustomIconProps {
  icon: CustomIconNames;
}
