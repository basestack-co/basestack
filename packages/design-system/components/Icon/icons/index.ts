export { default as GithubIcon } from "./GithubIcon";
export { default as GoogleColorsIcon } from "./GoogleColorsIcon";
export { default as GoogleIcon } from "./GoogleIcon";
