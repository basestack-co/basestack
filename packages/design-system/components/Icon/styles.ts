import styled from "styled-components";
import { space } from "styled-system";
import { rem } from "polished";

export const Container = styled.div.withConfig({
  shouldForwardProp: (prop) => !["size"].includes(prop),
})<{ size?: number }>`
  height: ${({ size }) => rem(`${size}px`)};
  width: ${({ size }) => rem(`${size}px`)};
  flex-shrink: 0;
  ${space};
`;

export const Icon = styled.span.withConfig({
  shouldForwardProp: (prop) => !["size", "muted", "color"].includes(prop),
})<{ size?: number; muted?: boolean; color?: string }>`
  color: ${({ theme, color, muted }) =>
    color || (muted ? theme.icon.muted : theme.icon.color)};
  font-size: ${({ size }) => rem(`${size}px`)};
`;
