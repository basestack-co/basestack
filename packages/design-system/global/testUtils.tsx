import { type RenderOptions, render } from "@testing-library/react";
import type React from "react";
// Styles
import { ThemeProvider } from "styled-components";
import theme from "../theme/lightTheme";

interface Props {
  children: React.ReactNode;
}

const WithThemeProvider = ({ children }: Props) => (
  <ThemeProvider theme={theme}>{children}</ThemeProvider>
);

const renderWithTheme = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, "wrapper">,
) => render(ui, { wrapper: WithThemeProvider, ...options });

export { renderWithTheme };
