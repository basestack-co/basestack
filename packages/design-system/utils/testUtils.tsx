import defaultTheme from "@basestack/design-system/theme/lightTheme";
import isPropValid from "@emotion/is-prop-valid";
import { RenderResult, render } from "@testing-library/react";
import React from "react";
import { StyleSheetManager, ThemeProvider } from "styled-components";

const Provider = ({
  children,
  theme = defaultTheme,
}: {
  children: React.ReactNode;
  theme: any;
}) => (
  <ThemeProvider theme={theme}>
    <StyleSheetManager shouldForwardProp={isPropValid}>
      {children}
    </StyleSheetManager>
  </ThemeProvider>
);

export const renderWithTheme = (
  ui: React.ReactElement,
  options?: any,
): RenderResult => render(ui, { wrapper: Provider, ...options });
