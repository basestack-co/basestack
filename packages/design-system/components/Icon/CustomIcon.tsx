import type React from "react";
import { useTheme } from "styled-components";
import { GithubIcon, GoogleColorsIcon, GoogleIcon } from "./icons";
import type {
  CustomIconCompProps,
  CustomIconNames,
  CustomIconProps,
} from "./types";

const iconMap: Record<CustomIconNames, React.ComponentType<CustomIconProps>> = {
  github: GithubIcon,
  google: GoogleIcon,
  google_colors: GoogleColorsIcon,
};

// Convert the keys of the iconMap into an array
export const iconNamesArray = Object.keys(iconMap) as Array<string>;

const CustomIcon = ({ icon, size, color }: CustomIconCompProps) => {
  const theme = useTheme();
  const IconComponent = iconMap[icon];

  if (!IconComponent) {
    return null;
  }

  return <IconComponent size={size} color={color || theme.icon.color} />;
};

export default CustomIcon;
