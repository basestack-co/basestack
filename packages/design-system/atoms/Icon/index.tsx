import React, { memo } from "react";
import { SpaceProps } from "styled-system";
import { Container, Icon } from "./styles";

export type Size = "xLarge" | "large" | "medium" | "small";

interface AvatarProps extends SpaceProps {
  /**
   * Icon name
   */
  icon: string;
  /**
   * Icon size
   */
  size?: Size;
  /**
   * Icon color
   */
  color?: string;
  /**
   * Icon muted color
   */
  muted?: boolean;
}

const handleSize = (size: Size) => {
  switch (size) {
    case "xLarge":
      return 48;
    case "large":
      return 32;
    default:
    case "medium":
      return 24;
    case "small":
      return 18;
  }
};

const Avatar = ({
  icon,
  size = "medium",
  color,
  muted,
  ...props
}: AvatarProps) => {
  const iconSize = handleSize(size);

  return (
    <Container size={iconSize} {...props}>
      <Icon
        data-testid="icon"
        color={color}
        muted={muted}
        size={iconSize}
        className="material-symbols-sharp"
      >
        {icon}
      </Icon>
    </Container>
  );
};

export default memo(Avatar);
