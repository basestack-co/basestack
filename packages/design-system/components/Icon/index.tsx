import { memo } from "react";
import CustomIcon, { iconNamesArray } from "./CustomIcon";
import { Container, Icon } from "./styles";
import { CustomIconNames, IconProps, Size } from "./types";

const handleSize = (size: Size) => {
  switch (size) {
    case "xxLarge":
      return 60;
    case "xLarge":
      return 48;
    case "large":
      return 32;
    default:
    case "medium":
      return 24;
    case "small":
      return 18;
  }
};

const IconComp = ({
  icon,
  size = "medium",
  color,
  muted,
  ...props
}: IconProps) => {
  const iconSize = handleSize(size);
  const isCustomIcon = iconNamesArray.includes(icon);

  return (
    <Container data-testid="icon-container" size={iconSize} {...props}>
      {isCustomIcon ? (
        <CustomIcon
          icon={icon as CustomIconNames}
          color={color}
          size={iconSize}
        />
      ) : (
        <Icon
          data-testid="icon"
          color={color}
          muted={muted}
          size={iconSize}
          className="material-symbols-rounded"
        >
          {icon}
        </Icon>
      )}
    </Container>
  );
};

export type { IconProps, Size, CustomIconNames };

export default memo(IconComp);
