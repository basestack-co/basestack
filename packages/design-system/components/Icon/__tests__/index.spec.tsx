import { cleanup } from "@testing-library/react";
import { renderWithTheme } from "../../../utils/testUtils";
import Icon from "..";

describe("Icon tests", () => {
  afterEach(cleanup);

  test("should render Icon correctly", () => {
    const { asFragment } = renderWithTheme(<Icon icon="help" />);
    expect(asFragment()).toMatchSnapshot();
  });

  test("should render muted Icon", () => {
    const { asFragment } = renderWithTheme(<Icon muted icon="help" />);
    expect(asFragment()).toMatchSnapshot();
  });

  test("should render red Icon", () => {
    const { asFragment } = renderWithTheme(<Icon color="red" icon="help" />);
    expect(asFragment()).toMatchSnapshot();
  });

  test("should render small Icon", () => {
    const { asFragment } = renderWithTheme(<Icon size="small" icon="help" />);
    expect(asFragment()).toMatchSnapshot();
  });

  test("should render medium Icon", () => {
    const { asFragment } = renderWithTheme(<Icon size="medium" icon="help" />);
    expect(asFragment()).toMatchSnapshot();
  });

  test("should render large Icon", () => {
    const { asFragment } = renderWithTheme(<Icon size="large" icon="help" />);
    expect(asFragment()).toMatchSnapshot();
  });

  test("should render xLarge Icon", () => {
    const { asFragment } = renderWithTheme(<Icon size="xLarge" icon="help" />);
    expect(asFragment()).toMatchSnapshot();
  });
});
