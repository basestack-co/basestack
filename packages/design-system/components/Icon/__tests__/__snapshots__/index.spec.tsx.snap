// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon tests render Icon correctly 1`] = `
<DocumentFragment>
  .c0 {
  height: 1.5rem;
  width: 1.5rem;
  flex-shrink: 0;
}

.c1 {
  color: #000000;
  font-size: 1.5rem;
}

<div
    class="c0"
    data-testid="icon-container"
    size="24"
  >
    <span
      class="c1 material-symbols-sharp"
      data-testid="icon"
      size="24"
    >
      help
    </span>
  </div>
</DocumentFragment>
`;
